PR: https://github.com/odoo/odoo/pull/246450

From: 84c1c564ba7389c61b9282e6d154bce4f0bfc75b
From: Shubham Agarwal(shag)
Date: 2026-04-24 12:58:59

Breaking data model changes scores: del:4 + add:0, change matches:
-    default_receipt_printer_id = fields.Many2one('pos.printer', string="Default Receipt Printer")
-    iface_cashdrawer = fields.Boolean(string='Cashdrawer', help="Automatically open the cashdrawer.")
-    pos_default_receipt_printer_id = fields.Many2one(related='pos_config_id.default_receipt_printer_id', readonly=False)
-    pos_iface_cashdrawer = fields.Boolean(string='Cashdrawer', compute='_compute_pos_iface_cashdrawer', readonly=False, store=True)

Total Changes: 38

[IMP] point_of_sale,**: printer management

**=pos_restaurant, pos_self_order

Following this commit:
* Removed default selection of printer from backend, instead given option in
frontend which solves multiple devices issue also.
* Moved iface_cashdrawer field from pos.config to pos.printer
* Remove `default_receipt_printer_id` field as it is no longer needed
* Popup that is been added in frontend will ask user to select default printer
option from the list of printers that are configured.
 > User can change the options from menu setting from Printers Section.

task-5494313

closes odoo/odoo#246450

Related: odoo/enterprise#105972
Related: odoo/upgrade#9381
Signed-off-by: Adrien Guilliams (adgu) <adgu@odoo.com>

================================= pseudo patch: =================================

--- a/addons/point_of_sale/models/pos_config.py
+++ b/addons/point_of_sale/models/pos_config.py
@@ -69,7 +69,6 @@ class PosConfig(models.Model):
     name = fields.Char(string='Point of Sale', required=True, help="An internal identification of the point of sale.")
     preparation_printer_ids = fields.Many2many('pos.printer', 'pos_config_printer_rel', 'config_id', 'printer_id', string="Preparation Printers", domain="[('use_type', '=', 'preparation')]")
     receipt_printer_ids = fields.Many2many('pos.printer', 'pos_config_receipt_printer_rel', 'config_id', 'printer_id', string="Receipt Printers", domain="[('use_type', '=', 'receipt')]")
-    default_receipt_printer_id = fields.Many2one('pos.printer', string="Default Receipt Printer")
     use_order_printer = fields.Boolean('Order Printer')
     is_installed_account_accountant = fields.Boolean(string="Is the Full Accounting Installed",
         compute="_compute_is_installed_account_accountant")
@@ -91,7 +90,6 @@ class PosConfig(models.Model):
     order_backend_seq_id = fields.Many2one('ir.sequence', string='Order Backend Sequence', readonly=True, copy=False)
     order_line_seq_id = fields.Many2one('ir.sequence', string='Order Line Sequence', readonly=True, copy=False)
     device_seq_id = fields.Many2one('ir.sequence', string='Device Sequence', readonly=True, copy=False)
-    iface_cashdrawer = fields.Boolean(string='Cashdrawer', help="Automatically open the cashdrawer.")
     iface_big_scrollbars = fields.Boolean('Large Scrollbars', help='For imprecise industrial touchscreens.')
     iface_group_by_categ = fields.Boolean("Group products by categories", help='Display products grouped by categories.')
     iface_print_auto = fields.Boolean(string='Automatic Receipt Printing', default=False,
@@ -212,12 +210,6 @@ class PosConfig(models.Model):
 
     pos_snooze_ids = fields.One2many('pos.product.template.snooze', 'pos_config_id', string='Snoozed Products')
 
-    @api.onchange('receipt_printer_ids')
-    def _onchange_receipt_printer_ids(self):
-        """Clear default_receipt_printer_id if it's removed from receipt_printer_ids"""
-        if self.default_receipt_printer_id.id not in self.receipt_printer_ids.ids:
-            self.default_receipt_printer_id = False
-
     def _get_next_order_refs(self, device_identifier='0'):
         next_number = self.order_backend_seq_id._next()
         year_2_digits = str(datetime.now().year)[-2:]

--- a/addons/point_of_sale/models/pos_printer.py
+++ b/addons/point_of_sale/models/pos_printer.py
@@ -65,6 +65,7 @@ class PosPrinter(models.Model):
         ),
     )
     use_lna = fields.Boolean(string="Use Local Network Access")
+    use_cashdrawer = fields.Boolean(string='Link Cashdrawer', help="Automatically open the cashdrawer.")
     paper_size = fields.Selection(string="Paper Size", selection=[
         ('80', 'Standard 80mm'),
         ('58', 'Standard 58mm'),
@@ -92,13 +93,20 @@ class PosPrinter(models.Model):
 
             record.paper_size_keys = ",".join(standard_size)
 
+    @api.onchange('use_type')
+    def _onchange_use_type(self):
+        """Disable use_cashdrawer when printer type is preparation"""
+        for rec in self:
+            if rec.use_type == "preparation" and rec.use_cashdrawer:
+                rec.use_cashdrawer = False
+
     @api.model
     def _load_pos_data_domain(self, data, config):
         return [('id', 'in', config.preparation_printer_ids.ids + config.receipt_printer_ids.ids)]
 
     @api.model
     def _load_pos_data_fields(self, config):
-        return ['id', 'name', 'product_categories_ids', 'printer_type', 'use_type', 'use_lna', 'printer_ip', 'paper_size', 'timeout']
+        return ['id', 'name', 'product_categories_ids', 'printer_type', 'use_type', 'use_lna', 'printer_ip', 'paper_size', 'use_cashdrawer', 'timeout']
 
     @api.constrains('printer_ip')
     def _constrains_printer_ip(self):

--- a/addons/point_of_sale/models/res_config_settings.py
+++ b/addons/point_of_sale/models/res_config_settings.py
@@ -54,7 +54,6 @@ class ResConfigSettings(models.TransientModel):
     pos_module_pos_restaurant = fields.Boolean(related='pos_config_id.module_pos_restaurant', readonly=False)
     pos_module_pos_appointment = fields.Boolean(related="pos_config_id.module_pos_appointment", readonly=False)
     pos_module_pos_avatax = fields.Boolean(related='pos_config_id.module_pos_avatax', readonly=False)
-    pos_default_receipt_printer_id = fields.Many2one(related='pos_config_id.default_receipt_printer_id', readonly=False)
     pos_receipt_printer_ids = fields.Many2many(related='pos_config_id.receipt_printer_ids', domain="[('use_type', '=', 'receipt')]", readonly=False)
     pos_preparation_printer_ids = fields.Many2many(related='pos_config_id.preparation_printer_ids', domain="[('use_type', '=', 'preparation')]", readonly=False)
     pos_use_custom_receipt_info = fields.Boolean(related='pos_config_id.use_custom_receipt_info', readonly=False)
@@ -82,7 +81,6 @@ class ResConfigSettings(models.TransientModel):
     pos_iface_available_categ_ids = fields.Many2many('pos.category', string='Available PoS Product Categories', compute='_compute_pos_iface_available_categ_ids', readonly=False, store=True)
     pos_iface_big_scrollbars = fields.Boolean(related='pos_config_id.iface_big_scrollbars', readonly=False)
     pos_iface_group_by_categ = fields.Boolean(related='pos_config_id.iface_group_by_categ', readonly=False)
-    pos_iface_cashdrawer = fields.Boolean(string='Cashdrawer', compute='_compute_pos_iface_cashdrawer', readonly=False, store=True)
     pos_iface_print_auto = fields.Boolean(related='pos_config_id.iface_print_auto', readonly=False)
     pos_iface_print_skip_screen = fields.Boolean(related='pos_config_id.iface_print_skip_screen', readonly=False)
     pos_iface_tax_included = fields.Selection(related='pos_config_id.iface_tax_included', readonly=False)
@@ -246,10 +244,6 @@ class ResConfigSettings(models.TransientModel):
             pos_config = self.env['pos.config'].browse(pos_config_id)
             return pos_config.open_ui()
 
-    @api.model
-    def _is_cashdrawer_displayed(self, res_config):
-        return res_config.pos_other_devices and bool(res_config.pos_default_receipt_printer_id)
-
     @api.depends('pos_limit_categories', 'pos_config_id')
     def _compute_pos_iface_available_categ_ids(self):
         for res_config in self:
@@ -266,14 +260,6 @@ class ResConfigSettings(models.TransientModel):
             else:
                 res_config.pos_selectable_categ_ids = self.env['pos.category'].search([])
 
-    @api.depends('pos_config_id', 'pos_default_receipt_printer_id', 'pos_other_devices')
-    def _compute_pos_iface_cashdrawer(self):
-        for res_config in self:
-            if self._is_cashdrawer_displayed(res_config):
-                res_config.pos_iface_cashdrawer = res_config.pos_config_id.iface_cashdrawer
-            else:
-                res_config.pos_iface_cashdrawer = False
-
     @api.depends('pos_use_header_or_footer', 'pos_config_id')
     def _compute_pos_receipt_header_footer(self):
         for res_config in self:
@@ -330,9 +316,3 @@ class ResConfigSettings(models.TransientModel):
                     old._add_trusted_config_id(config.pos_config_id)
                 if old.id in removed_trusted_configs:
                     old._remove_trusted_config_id(config.pos_config_id)
-
-    @api.onchange('pos_receipt_printer_ids')
-    def _onchange_pos_receipt_printer_ids(self):
-        """Clear pos_default_receipt_printer_id if it's removed from pos_receipt_printer_ids"""
-        if self.pos_default_receipt_printer_id.id not in self.pos_receipt_printer_ids.ids:
-            self.pos_default_receipt_printer_id = False
